File: packages/ts/src/core/history.ts
import type { HistoryEntry, HistoryAPI, WriteOperation, CacheSnapshot } from '../types.js';
import type { Cache } from './cache.js';

// ---------------------------------------------------------------------------
// HistoryLog — append-only, with a cursor for time travel
// ---------------------------------------------------------------------------

export type HistoryLog = HistoryAPI & {
  /** Record a new entry. Truncates any forward history after current cursor. */
  record(mutate: string, operation: WriteOperation, before: CacheSnapshot, after: CacheSnapshot): void;
  /** Current cursor index (points at the last applied entry). */
  cursor(): number;
};

export const createHistoryLog = (cache: Cache): HistoryLog => {
  const entries: HistoryEntry[] = [];
  let cursor = -1;

  const record = (
    mutate: string,
    operation: WriteOperation,
    before: CacheSnapshot,
    after: CacheSnapshot,
  ): void => {
    // Truncate any forward history (branching is not supported — linear only)
    if (cursor < entries.length - 1) {
      entries.splice(cursor + 1);
    }
    const entry: HistoryEntry = {
      index: entries.length,
      at: new Date().toISOString(),
      mutate,
      operation,
      before,
      after,
    };
    entries.push(entry);
    cursor = entries.length - 1;
  };

  const log = (): readonly HistoryEntry[] => [...entries];

  const back = (): void => {
    if (cursor < 0) return;
    const entry = entries[cursor];
    if (!entry) return;
    cache.restore(entry.before);
    cursor--;
  };

  const forward = (): void => {
    if (cursor >= entries.length - 1) return;
    cursor++;
    const entry = entries[cursor];
    if (!entry) return;
    cache.restore(entry.after);
  };

  const goto = (index: number): void => {
    if (index < 0 || index >= entries.length) return;
    const entry = entries[index];
    if (!entry) return;
    cache.restore(entry.after);
    cursor = index;
  };

  return {
    record,
    log,
    back,
    forward,
    goto,
    cursor: () => cursor,
  };
};

File: packages/swift/Sources/Antifragile/Core/History.swift
// StatelessUI — History
// Append-only log of every dispatched mutate. Supports back/forward/goto for
// time-travel debugging. Thread-safe via @MainActor (same actor as Cache + Store).

import Foundation
import Observation

@Observable
@MainActor
public final class HistoryLog {

    // MARK: - State
    private(set) public var entries: [HistoryEntry] = []
    private(set) public var cursor: Int = -1   // index of the "current" entry; -1 = empty

    // MARK: - Append

    /// Records a dispatched action. Truncates any forward entries (like a browser).
    public func append(_ entry: HistoryEntry) {
        // Drop any entries that were "in the future" relative to the cursor.
        if cursor < entries.count - 1 {
            entries = Array(entries[0...cursor])
        }
        entries.append(entry)
        cursor = entries.count - 1
    }

    // MARK: - Navigation

    /// Move the cursor one step back. Returns the entry moved to, or nil at start.
    @discardableResult
    public func back() -> HistoryEntry? {
        guard cursor > 0 else { return nil }
        cursor -= 1
        return entries[cursor]
    }

    /// Move the cursor one step forward. Returns the entry moved to, or nil at end.
    @discardableResult
    public func forward() -> HistoryEntry? {
        guard cursor < entries.count - 1 else { return nil }
        cursor += 1
        return entries[cursor]
    }

    /// Jump to an arbitrary index. Clamps to valid range.
    @discardableResult
    public func goto(index: Int) -> HistoryEntry? {
        guard !entries.isEmpty else { return nil }
        cursor = min(max(0, index), entries.count - 1)
        return entries[cursor]
    }

    // MARK: - Read

    /// Full log in chronological order.
    public func log() -> [HistoryEntry] { entries }

    /// The entry at the current cursor position.
    public var current: HistoryEntry? {
        guard cursor >= 0, cursor < entries.count else { return nil }
        return entries[cursor]
    }

    // MARK: - Reset

    public func clear() {
        entries = []
        cursor = -1
    }
}
